### PR TITLE
Added dim check for index_copy_

### PR DESCRIPTION
**Pull request by izdeby**
Fixing reported [bug](https://github.com/pytorch/pytorch/issues/20322)
The issue was related to not checking the dimensions of source vs destination tensors.

### CODE REVIEW

File: aten/src/ATen/native/Indexing.cpp [hunk 1]
```diff
@@ -272,7 +272,10 @@ Tensor & index_copy_(Tensor & self, int64_t dim, const Tensor & index, const Ten
   int64_t numIndices = index.numel();
   if (source.dim() == 0 && numIndices != 1) {
     AT_INDEX_ERROR("index_copy_(): When source is scalar, index should have one element (got ", numIndices, ")");
+  } else if (source.dim() != self.dim()) {
+    AT_INDEX_ERROR("index_copy_(): When source is not scalar, its dimention must match destinations");
```
Comment (VitalyFedyunin) [inline aten/src/ATen/native/Indexing.cpp]: /dimention/dimension/

Comment (VitalyFedyunin) [reply aten/src/ATen/native/Indexing.cpp]: Maybe: `When source is not scalar, its dimensionality must match the dimensionality of the destination`

**Review by VitalyFedyunin — DISMISSED** [review]

**Comment by VitalyFedyunin** [comment]
Test fail seems legit: `ERROR: test_index_copy_scalar_input_dim (__main__.TestAutograd)`

File: aten/src/ATen/native/Indexing.cpp [hunk 2]
```diff
@@ -272,7 +272,10 @@ Tensor & index_copy_(Tensor & self, int64_t dim, const Tensor & index, const Ten
   int64_t numIndices = index.numel();
   if (source.dim() == 0 && numIndices != 1) {
     AT_INDEX_ERROR("index_copy_(): When source is scalar, index should have one element (got ", numIndices, ")");
+  } else if (source.dim() != self.dim()) {
+    AT_INDEX_ERROR("index_copy_(): When source is not scalar, its dimensionality must match the dimensionality of the destination");
```
Comment (zou3519) [inline aten/src/ATen/native/Indexing.cpp]: It's nice to also list the dimensionalities so the user doesn't have to print out the tensors.
Something like:
```
index_copy_(): When source is not scalar, its dimensionality (3) must match the dimensionality of the destination (2)"
```

**Review by VitalyFedyunin — CHANGES_REQUESTED** [review]
Test fail seems legit: `ERROR: test_index_copy_scalar_input_dim (__main__.TestAutograd)`

**Review by gchanan — CHANGES_REQUESTED** [review]
test?

File: aten/src/ATen/native/Indexing.cpp (hunk 2 above)
Comment (gchanan) [reply aten/src/ATen/native/Indexing.cpp]: also, the error message doesn't match what is checked -- the message says source is not a scalar, but this isn't actually checked.

File: aten/src/ATen/native/Indexing.cpp [hunk 3]
```diff
@@ -263,16 +263,18 @@ Tensor & index_put_(Tensor & self, TensorList indices, const Tensor & value, con
 }
 
 Tensor & index_copy_(Tensor & self, int64_t dim, const Tensor & index, const Tensor & source) {
-  dim = maybe_wrap_dim(dim, self.dim());
```
Comment (VitalyFedyunin) [inline aten/src/ATen/native/Indexing.cpp]: Any reason to remove this wrapping? How you deal with negative values now?

**Review by VitalyFedyunin — APPROVED** [review]

**Comment by SsnL** [comment]
why does this not have a test?

**Pull request closed without merging**